api/projects/views.py

```diff
@@ -81,102 +86,77 @@ def permissions(self, *args, **kwargs):
 
     def get_queryset(self):
         if not self.kwargs.get("project_pk"):
             raise ValidationError("Missing project pk.")
 
-        return UserProjectPermission.objects.filter(
-            project__pk=self.kwargs["project_pk"]
-        )
-
-    def get_serializer_class(self):
-        if self.action == "list":
-            return ListUserProjectPermissionSerializer
-
-        return CreateUpdateUserProjectPermissionSerializer
+        return self.model_class.objects.filter(project__pk=self.kwargs["project_pk"])
 
     def perform_create(self, serializer):
-        project = Project.objects.get(id=self.kwargs["project_pk"])
-        serializer.save(project=project)
+        serializer.save(project_id=self.kwargs["project_pk"])
 
     def perform_update(self, serializer):
-        project = Project.objects.get(id=self.kwargs["project_pk"])
-        serializer.save(project=project)
+        serializer.save(project_id=self.kwargs["project_pk"])
 
 
-class UserPermissionGroupProjectPermissionsViewSet(viewsets.ModelViewSet):
-    pagination_class = None
-    permission_classes = [IsAuthenticated, NestedProjectPermissions]
+class UserProjectPermissionsViewSet(BaseProjectPermissionsViewSet):
+    model_class = UserProjectPermission
 
-    def get_queryset(self):
-        if not self.kwargs.get("project_pk"):
-            raise ValidationError("Missing project pk.")
+    def get_serializer_class(self):
+        if self.action == "list":
+            return ListUserProjectPermissionSerializer
 
-        return UserPermissionGroupProjectPermission.objects.filter(
-            project__pk=self.kwargs["project_pk"]
-        )
+        return CreateUpdateUserProjectPermissionSerializer
+
+
+class UserPermissionGroupProjectPermissionsViewSet(BaseProjectPermissionsViewSet):
+    model_class = UserPermissionGroupProjectPermission
 
     def get_serializer_class(self):
         if self.action == "list":
             return ListUserPermissionGroupProjectPermissionSerializer
 
         return CreateUpdateUserPermissionGroupProjectPermissionSerializer
 
-    def perform_create(self, serializer):
-        project = Project.objects.get(id=self.kwargs["project_pk"])
-        serializer.save(project=project)
 
-    def perform_update(self, serializer):
-        project = Project.objects.get(id=self.kwargs["project_pk"])
-        serializer.save(project=project)
+@swagger_auto_schema(method="GET", responses={200: UserObjectPermissionsSerializer()})
+@api_view(http_method_names=["GET"])
+@permission_classes([IsAuthenticated, IsProjectAdmin])
+def get_user_project_permissions(request, **kwargs):
+    user_id = kwargs["user_pk"]
+
+    project_permissions_calculator = ProjectPermissionsCalculator(kwargs["project_pk"])
+    user_permissions_data = (
+        project_permissions_calculator.get_user_project_permission_data(user_id)
+    )
+
+    # TODO: expose `user` and `groups` attributes from user_permissions_data
```

Comment:
This to-do can be removed, right?

Reply:
No, this is one that I added since I would like to, in the future expose these attributes in the API endpoint. The `UserProjectPermissionData` class has a `user` and `groups` attribute that contains information about where the permissions are coming from (i.e. from the user or from a group, and which group if so). 

As it's not necessary right now, I didn't want to spend the time to do it, however, I wanted to add the todo so that we know it can be done with the current functionality of the dataclass.